### ISSUE TITLE
Segfaults under docker buildx build --platform linux/amd64

### ISSUE THREAD

**Issue by benjamingwynn** [evanw/esbuild#3153]
The following minimal Dockerfile:

```dockerfile
FROM node
WORKDIR /tmp
RUN npm init -y
RUN npm i esbuild
```

Ran with the docker-buildx command from an aarch64 host (M1 Mac on Linux), like so:

```
docker buildx build --platform linux/amd64 -f <path to dockerfile above> .
```

Produces the following segfault:

```

#0 6.565 npm ERR! code 1
#0 6.566 npm ERR! path /tmp/node_modules/esbuild
#0 6.571 npm ERR! command failed
#0 6.572 npm ERR! command sh -c node install.js
#0 6.572 npm ERR! node:internal/errors:862
#0 6.573 npm ERR!   const err = new Error(message);
#0 6.573 npm ERR!               ^
#0 6.573 npm ERR!
#0 6.573 npm ERR! Error: Command failed: /tmp/node_modules/esbuild/bin/esbuild --version
#0 6.573 npm ERR! qemu: uncaught target signal 11 (Segmentation fault) - core dumped
#0 6.573 npm ERR!
#0 6.573 npm ERR!     at checkExecSyncError (node:child_process:885:11)
#0 6.574 npm ERR!     at Object.execFileSync (node:child_process:921:15)
#0 6.574 npm ERR!     at validateBinaryVersion (/tmp/node_modules/esbuild/install.js:98:28)
#0 6.574 npm ERR!     at /tmp/node_modules/esbuild/install.js:285:5 {
#0 6.574 npm ERR!   status: null,
#0 6.574 npm ERR!   signal: 'SIGSEGV',
#0 6.574 npm ERR!   output: [
#0 6.574 npm ERR!     null,
#0 6.575 npm ERR!     Buffer(0) [Uint8Array] [],
#0 6.575 npm ERR!     Buffer(67) [Uint8Array] [
#0 6.575 npm ERR!       113, 101, 109, 117,  58,  32, 117, 110,  99,  97, 117,
#0 6.575 npm ERR!       103, 104, 116,  32, 116,  97, 114, 103, 101, 116,  32,
#0 6.575 npm ERR!       115, 105, 103, 110,  97, 108,  32,  49,  49,  32,  40,
#0 6.575 npm ERR!        83, 101, 103, 109, 101, 110, 116,  97, 116, 105, 111,
#0 6.575 npm ERR!       110,  32, 102,  97, 117, 108, 116,  41,  32,  45,  32,
#0 6.575 npm ERR!        99, 111, 114, 101,  32, 100, 117, 109, 112, 101, 100,
#0 6.575 npm ERR!        10
#0 6.576 npm ERR!     ]
#0 6.576 npm ERR!   ],
#0 6.576 npm ERR!   pid: 35,
#0 6.576 npm ERR!   stdout: Buffer(0) [Uint8Array] [],
#0 6.576 npm ERR!   stderr: Buffer(67) [Uint8Array] [
#0 6.576 npm ERR!     113, 101, 109, 117,  58,  32, 117, 110,  99,  97, 117,
#0 6.576 npm ERR!     103, 104, 116,  32, 116,  97, 114, 103, 101, 116,  32,
#0 6.576 npm ERR!     115, 105, 103, 110,  97, 108,  32,  49,  49,  32,  40,
#0 6.576 npm ERR!      83, 101, 103, 109, 101, 110, 116,  97, 116, 105, 111,
#0 6.576 npm ERR!     110,  32, 102,  97, 117, 108, 116,  41,  32,  45,  32,
#0 6.577 npm ERR!      99, 111, 114, 101,  32, 100, 117, 109, 112, 101, 100,
#0 6.577 npm ERR!      10
#0 6.577 npm ERR!   ]
#0 6.577 npm ERR! }
```

**Comment by benjamingwynn** [comment]
Ah, the error coming from `qemu` should've given this away sooner, but it turns out you must start a buildx context before building with `--platform`. For example:


```sh
docker buiildx create --name foobar
docker buiildx use foobar

docker buildx build --platform linux/amd64 -f <path to dockerfile> .

docker buildx stop # cleanup the buildx context
```

Not sure why Docker doesn't warn you about this, but oh well.

Closing as this is not a bug with esbuild, leaving the issue around to help anyone who runs into this in the future.

**Comment by natevw** [comment]
I ran into this issue today, my issue/resolution wasn't *entirely* the same but this was very helpful if at the very least that it wasn't just me! Thanks much for these notes and links.

tl;dr; never got 100% to the bottom of it but Debian's own binfmt packages seem to be broken in a way that `tonistiigi/binfmt` fixes if you uninstall and then install that little bit of hopefully-trustworthy 3rd-party magic!

---

My setup is a Debian aarch64 VM running on an M1 macOS host and I was able to run most amd64 stuff just fine [e.g.](https://gist.github.com/natevw/9cd1c720398789a7dc5392d571058f7d) with the Qemu emulation but some indirect dependency on `esbuild` was just killing my whole actual app build. In my case I hadn't started with the third-party `--privileged` thing but had simply used `sudo apt install binfmt-support qemu-user-static` to get going.

I would use `docker buildx build --progress=plain --platform linux/amd64 -f Dockerfile-test -t esbt .` to build the following:

```
FROM docker.io/node:22-alpine

WORKDIR /app
RUN npm i esbuild --ignore-scripts
````

and then run it via `docker run -it --rm --platform=linux/amd64 esbt /bin/sh`, manually calling `node_modules/.bin/esbuild --version` from that shell. [Your way above is cleaner but I wanted it in a way to explore and troubleshoot more "hands-on".]

Anyway on Debian 11 (bullseye) with their own binfmt packages it would simply segfault outright (albeit with some intermediate node wrapper??):

```
/app # node_modules/.bin/esbuild --version
node:child_process:921
    throw err;
    ^

Error: Command failed: /app/node_modules/@esbuild/linux-x64/bin/esbuild --version
    at genericNodeError (node:internal/errors:983:15)
    at wrappedFn (node:internal/errors:537:14)
    at checkExecSyncError (node:child_process:882:11)
    at Object.execFileSync (node:child_process:918:15)
    at Object.<anonymous> (/app/node_modules/esbuild/bin/esbuild:221:28)
    at Module._compile (node:internal/modules/cjs/loader:1554:14)
    at Object..js (node:internal/modules/cjs/loader:1706:10)
    at Module.load (node:internal/modules/cjs/loader:1289:32)
    at Function._load (node:internal/modules/cjs/loader:1108:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14) {
  status: null,
  signal: 'SIGSEGV',
  output: [ null, null, null ],
  pid: 21,
  stdout: null,
  stderr: null
}

Node.js v22.14.0
```

Anyway on Debian 12 (bookworm) with the distro packages the `esbuild` command would actually log a version and *then* segfault (without the node stacktrace 🤷):

```
/app # node_modules/.bin/esbuild --version
0.25.2
Segmentation fault (core dumped)
```

Running just the `tonistiigi/binfmt` install above did not fix it, but running the uninstall **and then** the install did. Probably more clues in that Ubuntu bug report linked above but at this point I'd done enough yak shaving on the situation and kinda ready to just accept the magic container-escapee workaround.

**Comment by natevw** [comment]
Also due to [the way the `tonistiigi/binfmt` trick works](https://github.com/tonistiigi/binfmt/discussions/88) the re-configuration needs to be done every time the VM is restarted (confirmed at https://github.com/tonistiigi/binfmt/discussions/87 as well).


(I also messed around a bit with `multiarch/qemu-user-static` via https://github.com/docker/buildx/issues/314#issuecomment-2016366785 but whatever that's supposed to do didn't work on its own and the core `tonistiigi/binfmt` image is really sufficient afaict.)

**Comment by silverwind** [comment]
Still seeing this while running esbuild under `arm64` and `riscv64` emulation in QEMU, it is related to https://github.com/docker/buildx/issues/2028. This issue should be re-opened.
